x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -290,117 +264,72 @@ static String getPolicyId(SnapshotInfo snapshotInfo) {
                 " to have a policy in its metadata, but it did not"));
     }
 
-    /**
-     * Maybe delete the given snapshots. If a snapshot is currently running according to the cluster
-     * state, this waits (using a {@link ClusterStateObserver} until a cluster state with no running
-     * snapshots before executing the blocking
-     * {@link #deleteSnapshots(Map, TimeValue, SnapshotLifecycleStats)} request. At most, we wait
-     * for the maximum allowed deletion time before timing out waiting for a state with no
-     * running snapshots.
-     *
-     * It's possible the task may still run into a SnapshotInProgressException, if a snapshot is
-     * started between the state retrieved here and the actual deletion. Since is is expected to be
-     * a rare case, no special handling is present.
-     */
-    private void maybeDeleteSnapshots(Map<String, List<SnapshotInfo>> snapshotsToDelete,
-                                      TimeValue maximumTime,
-                                      SnapshotLifecycleStats slmStats) {
+    void deleteSnapshots(Map<String, List<SnapshotInfo>> snapshotsToDelete,
+                         SnapshotLifecycleStats slmStats,
+                         ActionListener<Void> listener) {
         int count = snapshotsToDelete.values().stream().mapToInt(List::size).sum();
         if (count == 0) {
             logger.debug("no snapshots are eligible for deletion");
             return;
         }
 
-        ClusterState state = clusterService.state();
-        if (okayToDeleteSnapshots(state)) {
-            logger.trace("there are no snapshots currently running, proceeding with snapshot deletion of [{}]",
-                formatSnapshots(snapshotsToDelete));
-            deleteSnapshots(snapshotsToDelete, maximumTime, slmStats);
-        } else {
-            logger.debug("a snapshot is currently running, rescheduling SLM retention for after snapshot has completed");
-            ClusterStateObserver observer = new ClusterStateObserver(clusterService, maximumTime, logger, threadPool.getThreadContext());
-            CountDownLatch latch = new CountDownLatch(1);
```

Comment:
All the other changes that make methods async by adding `ActionListener`s come from removing the blocking here. I tried to not make any logical changes outside of moving to async and to keep those changes purely mechanical.